### ISSUE TITLE
Make events respond to individual event insertions/deletions

### ISSUE THREAD

**Issue by DanChadwick** [scoutforpets/ember-fullcalendar#16]
The work in PR #9 was reverted. There is a lot of good work there, and I think the idea is sound -- to cause fullcalendar to mirror changes to events array without removing and re-adding all the events, with the performance and visual consequences.

The PR is closed because it was merged into branch `upstream/miguelcobain-view-change`. That branch was abandoned.

In #9 there was discussion of the change breaking code which uses an events function. I think all that is needed is a few tests or a different attribute to distinguish from these two overloaded uses. `events` can be:

- [an array](http://fullcalendar.io/docs/event_data/events_array/)
- [the URL of a json feed](http://fullcalendar.io/docs/event_data/events_json_feed/) or
- [a function](http://fullcalendar.io/docs/event_data/events_function/)

So I guess the questions is whether there is interest in pushing @miguelcobain's code forward so that it would be compatible with the above.

**Comment by jamesdixon** [comment]
@DanChadwick when we merged that PR, we ended up having a number of issues, hence the reason we decided to abandon it. While I agree that the concept of observing the events array directly appears to be a good idea, we ran into a handful of issues that led us to revert to a simpler method of observing the array. 

There was one issue specifically that resulted in that decision (forgive me for not recalling all of the specifics):

 [The way in which inserted events were being processed](https://github.com/miguelcobain/ember-fullcalendar/commit/ea62b9845856b4b3fcc51971c806af984f19ec89#diff-5c38c38a49015ad83f7d342954334047L277) was causing duplicates when using the `eventDataTransform` method to format incoming events. When looking at Miguel's repo, I discovered that there was [another commit](https://github.com/miguelcobain/ember-fullcalendar/commit/ea62b9845856b4b3fcc51971c806af984f19ec89#diff-5c38c38a49015ad83f7d342954334047R275) not included in the PR that replaced the aforementioned problematic code. When switching out that method, performance was extremely poor. Switching to the simpler method of calling the `rerender` method when the array changes fixed both the issue of duplicate events as well as putting restoring an acceptable level of performance.

While I can't confirm this, I had read that FC only re-renders events that have changed. I don't have the source offhand or know for sure if it's correct, but something I thought I'd mention.

While I like the concept of only re-rendering what's changed, I just needed to move on and get work done on the project in which this is being used. 

That said, if you'd like to revisit this, you're more than welcome to do so.

One other thing: I'm no Ember expert by any means, but in the grand scheme of things, are we gaining anything by adding a handful of new observers to make this happen? I only ask because I always hear that we should stay away from observers when possible.

**Comment by DanChadwick** [comment]
First, thank you for your work on this. I hope that I didn't unintentionally imply anything other than gratitude toward this addon's maintainers or @miguelcobain. 

I'm hoping @miguelcobain can comment more on the issues you had, including performance.

With regard to the observer, both solutions have the same number of observers really. When you observe `events.[]` you are firing that observer every time the array change or elements are added or removed. When you observe `events`, you are only triggering the observer when the entire array object itself changes. Depending upon how the addon is used, this might well be never.

But then there is the added impact of using the array observer. If the consuming app adds/deletes lots of events one at a time I can see that this might be very expensive. I'm also not sure how updating is handled. Hmmm.

And last there is the impact of wrapping the objects. There could be an impact there.

I don't have a solution, but maybe it's worth a bit more discussion if @miguelcobain is up to it. I believe he is using his solution in production. Maybe there is a difference of scale in the number of events that need to be supported.

Again, thanks for all your efforts.

**Comment by jamesdixon** [comment]
No offense taken to anything you had said whatsoever 😄  I just wanted to be clear as to why we ended up reverting the PR. In hindsight, I probably should have reached out to @miguelcobain to discuss further and come up with a solution, but as I had mentioned, we were spending a lot of time trying to get it to "fit" and needed to switch gears and make progress on the project and come back to the addon later.

As for number of supported events, the project I'm working on needs to handle lots. In some cases there may be over 100 events per day. While performance is acceptable at this point, there is definitely room for improvement and we're seeing a lot of that time spent in the rendering of those events. It may be more of a FullCalendar issue than something we can solve directly. I came across [this issue](https://github.com/fullcalendar/fullcalendar/issues/2524) that suggests a fairly major regression in performance with FC 2.x compared to 1.x. If you Google "fullcalendar performance," you'll see a ton of issues regarding rendering performance.

Thank you for all of _your_ help in making this addon better. Really appreciate the efforts of you and @miguelcobain
